### ISSUE TITLE
Custom formatter to output the number of errors found

### ISSUE THREAD

**Issue by s10wen** [stylelint/stylelint#659]
I'm wondering if this is possible please?

I'm using the `cli` and `stylelint **/*.scss --syntax scss` is working great, but I'd like it to also output, at the end something like this;

```
412: (property-no-vendor-prefix) found
203: (rule-trailing-semicolon) found
```

(where 412 and 203 are the amount of errors that were found)

To give this some context, I'm working on a very large code base that had been added to over time by a number of developers and I'm looking to tidy it all up. So it would be nice if I could log how many errors there were to begin with and I can see improvements being made over time, by getting the error numbers down.

**Comment by jeddy3** [comment]
> I'm wondering if this is possible please?

Absolutely. Sounds like a great idea for a custom formatter. There is some information [in the docs](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/formatters.md) that you'll need if you want to give it a go.

There are also two custom formatter that you can harvest bits from:
https://github.com/jeddy3/stylelint-formatter-yhw
https://github.com/stylelint/stylelint-checkstyle-formatter

The `warnings` array in the _stylelint result objects_ contains the rule name. By counting the occurrences of these you'll be able to tally up a total for each rule.

Good luck :)

@davidtheclark What's the naming convention btw? Are we going with `stylelint-x-formatter` or `stylelint-formatter-x`. I think my buddy Barnaby is the only person who uses `stylelint-formatter-yhw` and so happy to rename it to `stylelint-yhw-formatter` for the sake on consistency.

**Comment by s10wen** [comment]
I'm not familiar with all those things, but I managed to get it working using the `cli` with;

```
stylelint /**/*.scss --syntax scss > lint.txt
```

The `> lint.txt` saves the output to a text file, then from that file did some bash:

```
grep -o "(property-no-vendor-prefix)" lint.txt | wc -l | xargs echo "(property-no-vendor-prefix):" >> together.txt
grep -o "(rule-trailing-semicolon)"   lint.txt | wc -l | xargs echo "(rule-trailing-semicolon):"   >> together.txt
```

Does the job for what I'm after for now, but would be nice if there was a better solution :)

**Comment by davidtheclark** [comment]
@jeddy3: I'm fine with either way. We don't have an existing convention, I believe: plugins just start with `stylelint-*`. So anything starting with `stylelint-*` seems fine.

@s10wen : Nice CLI solution. But as @jeddy3 said, a custom formatter is the way to go.

**Comment by davidtheclark** [comment]
Oh, sorry, @jeddy3 I realize that you changed the title so this might spur someone to make that formatter.

**Comment by jeddy3** [comment]
I did indeed :)

**Comment by bluetidepro** [comment]
@jeddy3 thought it would be a good idea to help share my task with the custom stuff with the formatter.

@s10wen you can see a snippet of my code here: https://github.com/stylelint/stylelint/issues/400#issuecomment-176782869 and I can show you below how the whole picture works, the full grunt task (*if this helps*):

````js
/* jshint node:true */
'use strict';

var chalk = require('chalk');
var symbols = require('log-symbols');
var path = require('path');
var grunt = require('grunt');
var postcss = require('postcss');
var _ = require('lodash');
var scsssyntax = require('postcss-scss');
var stylelint = require('stylelint');
var reporter = require('postcss-reporter');

// Used in reporterMessagesFormat as a regex
var oldCodeRegexTest = /old\/components\/(.+)/;

var files = grunt.file.expand([
	'scss/**/*.scss',
]);

function reporterLogFrom(fromValue) {
	if (fromValue.charAt(0) === '<') return fromValue;
	return path.relative(process.cwd(), fromValue);
}

function reporterMessagesFormat(message, source) {
    // This is a regex test we do so that certain .scss/css in certain folders is considered
    // an error while old code we haven't migrated yet is just a warning
	var isOldCode = oldCodeRegexTest.test(source);
	var str = '';
	if (message.text) {
        // Just get the message text
		var regExp = /(.+)( \(.+\))/; 
		var messageText = regExp.exec(message.text);
        // Get the column/line info
		if (message.line && message.column) {
			str += chalk.gray(message.line + ':' + message.column + ' ');
		}
        // Check whether we should call this an error or not
		if (message.severity === 'warning' && !isOldCode) {
			str += chalk.yellow(symbols.warning + ' Warning:');
		} else if(message.severity === 'error' || isOldCode) {
			str += chalk.red(symbols.error + ' Error:');
		}
        // Link to the rule so in apps like iTerm they can cmd click the link to open
        // to that rule in the browser
		str += ' ' + messageText[1];
		if (message.rule) {
			str += chalk.gray(' (' + chalk.blue('http://tiny.cc/pxoy7x/' + message.rule) + ')');
		}
	}
	return str;
}

module.exports = function(grunt) {
	grunt.registerTask('scsslint', 'Lint CSS', function() {
		var finished = this.async();
		var totalProcessedFiles = 0;
		var totalErrors = [];
		var totalWarnings = [];
		var successFiles = [];

		// console.log(files);
		console.log(); // empty line
		// console.log('Files.length: ' + files.length + '\n');

		files.forEach(function(file) {
			var fileContents = grunt.file.read(file);
			var stylelintPostcss = postcss([
				stylelint(),
				reporter({
					formatter: function(input) {
						// Custom formatter to display the warnings/errors
						var messages = input.messages;
						var output = '';

           				// If there is a source, output it
						if (input.source) {
							output += chalk.underline(reporterLogFrom(input.source) + '\n');
						}

						// For each message (with content) loop the function `reporterMessagesFormat`, so lets customize how that output looks
						messages.forEach(function(message) {
							output += reporterMessagesFormat(message, input.source) + '\n';
						})

						// Ignore messages that are empty
						if(!_.isEmpty(messages)){
							return output;
						}
					},
					throwError: false
				})
			]);

			postcss().process(fileContents, {
				parser: scsssyntax,
				from: file
			}).then(function (cleaned) {
			    return stylelintPostcss.process(cleaned.css, {
					parser: scsssyntax,
					from: file
				});
			}).then(function(result) {
				totalProcessedFiles++;

				// Lets run through the messages and display some totals
				if(result.messages.length){
					var isOldCode = oldCodeRegexTest.test(result.opts.from);
					if(isOldCode){
						result.messages.map(function(message){
						    message.severity = 'error';
						    return message;
						});
					}
					totalErrors.push(_.filter(result.messages, { 'severity': 'error' }).length);
					totalWarnings.push(_.filter(result.messages, { 'severity': 'warning' }).length);
				} else {
					successFiles.push(result.opts.from);
				}

				// Everything is done, lets recap and finish
				if (totalProcessedFiles === files.length) {

					// Tell you about the success files
					if(successFiles.length){
						var listSuccessFiles = '';
						successFiles.forEach(function(i, idx, array){
						   if (idx === array.length - 1){
								listSuccessFiles += 'and "' + i + '".';
						   } else {
								listSuccessFiles += '"' + i + '", ';
						   }
						});
						grunt.verbose.writeln(
							chalk.underline.green(symbols.success + '  Sweet. 0 total .scss errors found in files: \n') + chalk.italic(listSuccessFiles)
						);
					}

					// Tell you about the error/warning files
					totalErrors = eval(totalErrors.join('+'));
					totalWarnings = eval(totalWarnings.join('+'));
					if(totalErrors || totalWarnings){
						if(totalErrors > 0){
							grunt.log.writeln(
								chalk.red(symbols.error + '  ' + totalErrors + ' .scss total error(s) total. Well, shit.')
							);
							finished(false);
						}
						if(totalWarnings > 0){
							grunt.log.writeln(
								chalk.yellow(symbols.warning + '  ' + totalWarnings + ' .scss total warnings(s) total.')
							);
							finished();
						}
					} else {
						grunt.log.writeln(
							chalk.green(symbols.success + '  ' + 'No .scss errors!')
						);
						finished();
					}
				}
			});

		});
	});
};
````

This isn't the most clean and elegant way to do it, I'm sure, but it works great for my use case! :)

**Comment by jeddy3** [comment]
Thanks for this @bluetidepro :)

**Comment by s10wen** [comment]
@bluetidepro cool, I might have a look at a `gulp` solution as well, still playing about and thinking about what my end goals are for the code base I'm working on will be.

**Comment by jeddy3** [comment]
Just doing some issue triage...

Closing this in favour of tracking #703

**Comment by kylezeeuwen** [comment]
I wrote the best grep! You guys are going to love it it's fantastic.

    stylelint **/*.less | grep xpected | awk '{print $(NF-1),"\n",$NF}' | tr -d ' ' | grep -v 'line(s)' | grep -v '"' | grep -v "'" | sort | uniq -c | sort -n

Heres what the output looks like:

       1 at-rule-name-newline-after
       1 comment-empty-line-before
       1 declaration-block-no-duplicate-properties
    ...
     712 block-closing-brace-space-after
    1190 block-closing-brace-space-before
    1361 block-opening-brace-space-after
    4500

The only issue I have is that final number (4500) is not the sum (that would be nice) instead it is a count of lines that didn't end with the rule but ended with some kinda white spacey thing.

If you want to know the total count of errors, note the count of white spacey things (in my case 4500) , then replace <NUMBER> with the number in the command below:

    grunt stylelint | grep xpected | awk '{print $(NF-1),"\n",$NF}' | tr -d ' ' | grep -v 'line(s)' | grep -v '"' | grep -v "'" | sort | uniq -c | sort -n | grep -v <NUMBER> | sed -e 's/^ *//g' | cut -d ' ' -f 1 | paste -sd+ - | bc
 
    For example with 4500:
    grunt stylelint | grep xpected | awk '{print $(NF-1),"\n",$NF}' | tr -d ' ' | grep -v 'line(s)' | grep -v '"' | grep -v "'" | sort | uniq -c | sort -n | grep -v 4500 | sed -e 's/^ *//g' | cut -d ' ' -f 1 | paste -sd+ - | bc


Bam, error count:

4995

I clearly have some work to do ...

HTH

**Comment by jeddy3** [comment]
@kylezeeuwen That is a rather nice grep.

FYI, a `verbose` [formatter](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/node-api.md#formatter) was added to stylelint in the last year, which tallies up the number of sources checked, the total number of warnings and the number of warnings _per rule_.  The ouput looks like this:
 
<img width="739" alt="screen shot 2016-05-02 at 22 01 15" src="https://cloud.githubusercontent.com/assets/808227/14970073/0ebc8bd4-10be-11e6-9398-4aafa0657fa4.png">

You can use it as so:

```shell
stylelint "**/*.less" --formatter verbose
```

> I clearly have some work to do ...

Good luck! Perhaps [stylefmt](https://github.com/morishitter/stylefmt) can help with a few of the code style warnings.
